### PR TITLE
Support VPPFhenBInBalancedMemory to enhance memory efficient FthenB

### PR DESCRIPTION
**Pull request by zhangyuqin1998**
### PR Category

Distributed Strategy

### PR Types

New features

### Description

Enhancing memory efficient with VPPFhenBInBalancedMemory by releasing activations earlier when utilizing the FthenB scheduling mode in pipeline parallelism.

开启方法：复用best_unbalanced_scheduler开关
``` yaml
"pipeline_parallel_config": "best_unbalanced_scheduler",
```

示意图：

4f728cff067b9e72686429e120b0707e

### CODE REVIEW

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 1]
```diff
@@ -2351,3 +2351,260 @@ def forward_backward_pipeline(
 
         self.timer_printer()
         return train_loss
+
+
+class PipelineParallelWithInterleaveFthenBMemoryEfficient(
```
Comment (ForFishes) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: The class name is too long, simplify it

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 2]
```diff
@@ -2351,3 +2351,260 @@ def forward_backward_pipeline(
...
+            self.input_tensors[next_forward_virtual_pp_rank].append(
+                input_tensor
+            )
+            self._release_output(output_tensor)
+
+        if self.is_pipeline_first_stage(ignore_virtual=True):
+            assert (
+                forward_send_recv_buffer_queue.qsize()
+                == num_steps - startup_steps - 1
+            ), forward_send_recv_buffer_queue.qsize()
```
Comment (ForFishes) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:2464]: self.accumulate_steps - (self.num_stages - 1) ??

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py [hunk 3]
```diff
@@ -2351,3 +2351,260 @@ def forward_backward_pipeline(
...
+        logger.info("Using PipelineParallelWithInterleaveFthenBMemoryEfficient")
+
+    def forward_backward_pipeline(
+        self,
+        data,
+        scaler,
+        forward_only=False,
+        compute_loss=True,
+        return_micro_batch_loss=False,
+    ):
```
Comment (ForFishes) [inline python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:2368]: If you need to support forward_only, you can simply let it go to the normal branch.

File: python/paddle/distributed/fleet/model.py [hunk 4]
```diff
@@ -167,9 +168,16 @@ def distributed_model(model):
                     model, fleet_env._hcg, strategy=strategy
                 )
             elif pp_degree <= accumulate_steps < 2 * pp_degree:
-                model = PipelineParallelWithInterleaveFthenB(
-                    model, fleet_env._hcg, strategy=strategy
-                )
+                if strategy.hybrid_configs[
+                    "pp_configs"
+                ].best_unbalanced_scheduler:
+                    model = PipelineParallelWithInterleaveFthenBMemoryEfficient(
```
Comment (ForFishes) [inline python/paddle/distributed/fleet/model.py]: overlap is not support currently， using assert

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 2 above)
Comment (zhangyuqin1998) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:2464]: > self.accumulate_steps - (self.num_stages - 1) ??

是 self.accumulate_steps - (self.num_stages - self.stage_id)，但感觉还是写成 num_steps - startup_steps - 1 更好理解～

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 1 above)
Comment (zhangyuqin1998) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py]: > The class name is too long, simplify it

done

File: python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py (hunk 3 above)
Comment (zhangyuqin1998) [reply python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:2368]: > If you need to support forward_only, you can simply let it go to the normal branch.

done

**Review by ForFishes — APPROVED** [review]
LGTM

**Pull request merged**